File: mobile/lib/api/api_client.dart
import 'package:dio/dio.dart';
import 'package:flutter/material.dart';
import 'package:syspig/services/prefs_service.dart';

class ApiClient {
  late Dio _dio;

  ApiClient() {
    BaseOptions options = BaseOptions(
      baseUrl: 'http://localhost:3000',
    );

    _dio = Dio(options);

    bool isRefreshing = false;

    // Adiciona interceptors imediatamente
    _dio.interceptors.add(InterceptorsWrapper(
      onRequest: (options, handler) async {
        String? accessToken = await PrefsService.getAccessToken();
        int? userId = await PrefsService.getUserId();
        int? fazendaId = await PrefsService.getFazendaId();

        if (accessToken != null) {
          options.headers['Authorization'] = 'Bearer $accessToken';
        }
        if (userId != null) {
          options.headers['user-id'] = userId.toString();
        }
        if (fazendaId != null) {
          options.headers['fazenda-id'] = fazendaId.toString();
        }

        return handler.next(options);
      },
      onError: (error, handler) async {
        String? refreshToken = await PrefsService.getRefreshToken();
        if (error.response?.statusCode == 403 && refreshToken != null) {

          // Evita loops definindo um flag
          if (isRefreshing) {
            return handler.next(error); // Já foi tentado, retorna o erro original
          }

          isRefreshing = true;

          // Marca como tentativa de renovação
          error.requestOptions.extra["retry"] = true;

          // Tentativa de renovar o accessToken com o refreshToken
          try {
            var response = await _dio.post('/auth/refresh', data: {
              'refreshToken': refreshToken,
            });

            if (response.statusCode == 200) {
              // Armazena o novo accessToken
              String newAccessToken = response.data['accessToken'];
              await PrefsService.saveAccessToken(newAccessToken);

              // Repete a requisição original com o novo accessToken
              error.requestOptions.headers['Authorization'] = 'Bearer $newAccessToken';
              
              final retryResponse = await _dio.request(
                error.requestOptions.path,
                options: Options(
                  method: error.requestOptions.method,
                  headers: error.requestOptions.headers,
                ),
                data: error.requestOptions.data,
                queryParameters: error.requestOptions.queryParameters,
              );
              return handler.resolve(retryResponse);
            }
          } catch (e) {
            // Se a renovação falhar, verifica se o erro é devido ao refreshToken expirado
            if (e is DioException && e.response?.statusCode == 403) {
              await PrefsService.logout(); // Limpa os dados locais
              return;
            }

            // Retorna o erro original em caso de falha geral
            return handler.reject(
              DioException(
                requestOptions: error.requestOptions,
                error: e,
                type: DioExceptionType.unknown,
              ),
            );
          }
        }

        return handler.next(error);
      },
    ));
  }

  Dio get dio => _dio;
}

File: web/src/apiClient.ts
import axios from 'axios';
import Cookie from 'js-cookie';

// Criação de um cliente Axios
const apiClient = axios.create({
  baseURL: process.env.NEXT_PUBLIC_API_URL,
  headers: {
    'Content-Type': 'application/json',
  },
});

// Adicionar interceptador de requisição
apiClient.interceptors.request.use(
  async (config) => {
    // Buscar o accessToken do cookie
    const accessToken = Cookie.get('accessToken');
    const userId = Cookie.get('userId');
    const fazendaId = Cookie.get('fazendaId');

    if (accessToken) {
      config.headers['Authorization'] = `Bearer ${accessToken}`;
    }

    if (userId) {
      config.headers['user-id'] = userId;
    }

    if (fazendaId) {
      config.headers['fazenda-id'] = fazendaId;
    }

    return config;
  },
  (error) => {
    return Promise.reject(error);
  }
);

// Adicionar interceptador de resposta para lidar com a renovação de token
apiClient.interceptors.response.use(
  (response) => response,
  async (error) => {
    // Se o erro for 403, tentar renovar o token
    if (error.response?.status === 403) {
      const refreshToken = Cookie.get('refreshToken');
      
      if (refreshToken) {
        try {
          // Tentar renovar o accessToken com o refreshToken
          const refreshResponse = await axios.post(`${process.env.NEXT_PUBLIC_API_URL}/auth/refresh`, {
            refreshToken,
          });

          const newAccessToken = refreshResponse.data.accessToken;
          Cookie.set('accessToken', newAccessToken, { expires: 1 });

          // Retentar a requisição original com o novo token
          error.config.headers['Authorization'] = `Bearer ${newAccessToken}`;
          return axios(error.config);
        } catch (refreshError) {
          // Se falhar, limpar os cookies e redirecionar para login
          Cookie.remove('accessToken');
          Cookie.remove('refreshToken');
          window.location.href = '/login';
          return Promise.reject(refreshError);
        }
      }
    }

    return Promise.reject(error);
  }
);

export default apiClient;
